File: types/return-order.go
package types

import "github.com/johnlettman/oyster/util"

// ReturnOrder represents the priority of returns for the LIDAR to output.
// The LIDAR can have more than 1 or 2 detected "returns."
// ReturnOrder indicates to the LiDAR which ones it should output.
//
// For additional information, refer to [return_order].
//
// [return_order]: https://static.ouster.dev/sensor-docs/image_route1/image_route2/common_sections/API/sensor_configuration_description.html?highlight=return_order#return-order)
type ReturnOrder int

const (
	ReturnOrderUnspecified        ReturnOrder = iota
	ReturnOrderStrongestToWeakest             // The priority of LiDAR returns with the strongest returns first.
	ReturnOrderFarthestToNearest              // The priority of LiDAR returns with the farthest returns first.
	ReturnOrderNearestToFarthest              // The priority of LiDAR returns with the nearest returns first.

	// Deprecated: Only present in old test firmware.
	ReturnOrderStrongestReturnFirst // The priority of LiDAR returns with the strongest returns first.

	// Deprecated: Only present in old test firmware.
	ReturnOrderLastReturnFirst // The priority of LiDAR returns with the last returns first.
)

var (
	// returnOrderStringKV maps ReturnOrder values to their string representations.
	returnOrderStringKV = map[ReturnOrder]string{
		ReturnOrderUnspecified:          "unspecified",
		ReturnOrderStrongestToWeakest:   "strongest to weakest",
		ReturnOrderNearestToFarthest:    "nearest to farthest",
		ReturnOrderFarthestToNearest:    "farthest to nearest",
		ReturnOrderStrongestReturnFirst: "strongest return first",
		ReturnOrderLastReturnFirst:      "last return first",
	}

	// returnOrderGoStringKV maps ReturnOrder values to their Go syntax representations.
	returnOrderGoStringKV = map[ReturnOrder]string{
		ReturnOrderUnspecified:          "ReturnOrderUnspecified",
		ReturnOrderStrongestToWeakest:   "ReturnOrderStrongestToWeakest",
		ReturnOrderNearestToFarthest:    "ReturnOrderNearestToFarthest",
		ReturnOrderFarthestToNearest:    "ReturnOrderFarthestToNearest",
		ReturnOrderStrongestReturnFirst: "ReturnOrderStrongestReturnFirst",
		ReturnOrderLastReturnFirst:      "ReturnOrderLastReturnFirst",
	}

	// returnOrderTextKV maps ReturnOrder values to their text representations.
	returnOrderTextKV = map[ReturnOrder]string{
		ReturnOrderUnspecified:          "UNSPECIFIED",
		ReturnOrderStrongestToWeakest:   "STRONGEST_TO_WEAKEST",
		ReturnOrderNearestToFarthest:    "NEAREST_TO_FARTHEST",
		ReturnOrderFarthestToNearest:    "FARTHEST_TO_NEAREST",
		ReturnOrderStrongestReturnFirst: "STRONGEST_RETURN_FIRST",
		ReturnOrderLastReturnFirst:      "LAST_RETURN_FIRST",
	}

	// returnOrderTextVK maps string representations to ReturnOrder values.
	returnOrderTextVK = util.ReverseMap(returnOrderTextKV)
)

// String returns the string representation of an ReturnOrder value.
func (r ReturnOrder) String() string {
	if s, ok := returnOrderStringKV[r]; ok {
		return s
	}

	return returnOrderStringKV[ReturnOrderUnspecified]
}

// GoString returns the Go syntax representation of an ReturnOrder value.
func (r ReturnOrder) GoString() string {
	if s, ok := returnOrderGoStringKV[r]; ok {
		return s
	}

	return returnOrderGoStringKV[ReturnOrderUnspecified]
}

// MarshalText returns the text representation of an ReturnOrder value.
func (r ReturnOrder) MarshalText() ([]byte, error) {
	if text, ok := returnOrderTextKV[r]; ok {
		return []byte(text), nil
	}

	return []byte(returnOrderTextKV[ReturnOrderUnspecified]), nil
}

// UnmarshalText updates the value of ReturnOrder by unmarshalling the given text.
// It checks if the text matches any string representation in returnOrderTextVK.
//   - If a match is found, the corresponding ReturnOrder value is assigned to r.
//   - Otherwise, ReturnOrderUnspecified is assigned to r.
//
// UnmarshalText always returns nil as error.
func (r *ReturnOrder) UnmarshalText(text []byte) error {
	if returnOrder, ok := returnOrderTextVK[string(text)]; ok {
		*r = returnOrder
	} else {
		*r = ReturnOrderUnspecified
	}

	return nil
}

File: src/types/return_order.rs
use std::fmt::{Display, Formatter};
#[allow(clippy::deprecated, deprecated)]
use serde::{Deserialize, Serialize};

/// Represents the priority of returns for the LIDAR to output.
/// The LIDAR can have more than 1 or 2 detected "returns."
/// This indicates to the LIDAR which ones it should output.
/// For additional information, refer to [`return_order`].
///
/// [`return_order`]: https://static.ouster.dev/sensor-docs/image_route1/image_route2/common_sections/API/sensor_configuration_description.html?highlight=return_order#return-order
#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Copy, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ReturnOrder {
    /// The priority of LIDAR returns with the strongest returns first.
    StrongestToWeakest,

    /// The priority of LIDAR returns with the farthest returns first.
    NearestToFarthest,

    /// The priority of LIDAR returns with the nearest returns first.
    FarthestToNearest,

    /// The priority of LIDAR returns with the strongest returns first.
    #[deprecated = "Only present in old test firmware."]
    StrongestReturnFirst,

    /// The priority of LIDAR returns with the last returns first.
    #[deprecated = "Only present in old test firmware."]
    LastReturnFirst,
}

impl Default for ReturnOrder {
    fn default() -> Self {
        Self::StrongestToWeakest
    }
}

impl Display for ReturnOrder {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", match *self {
            Self::StrongestToWeakest => "strongest to weakest",
            Self::NearestToFarthest => "nearest to farthest",
            Self::FarthestToNearest => "farthest to nearest",
            Self::StrongestReturnFirst => "strongest return first",
            Self::LastReturnFirst => "last return first"
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use test_log::test;
    use log::info;

    #[test]
    fn test_default() {
        let want = ReturnOrder::StrongestToWeakest;
        let got = ReturnOrder::default();
        assert_eq!(want, got);
    }

    #[test]
    fn test_display() {
        let cases = vec![
            (ReturnOrder::StrongestToWeakest, "strongest to weakest"),
            (ReturnOrder::NearestToFarthest, "nearest to farthest"),
            (ReturnOrder::FarthestToNearest, "farthest to nearest"),
            (ReturnOrder::StrongestReturnFirst, "strongest return first"),
            (ReturnOrder::LastReturnFirst, "last return first")
        ];

        for (return_order, want) in cases {
            info!("Displaying {return_order:?}, expecting {want:?}");
            let got = format!("{return_order}");
            assert_eq!(want, got);
        }
    }
}
